### PR TITLE
Straighten up MappedOperator hierarchy and typing

### PR DESCRIPTION
**Pull request by uranusjr**
This PR may be too large.

~~Details to come later. Submitted for CI at the moment.~~

After some discussion with @ashb, this simplfies the operator mapping API to implement only what’s described in [AIP-42](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-42+Dynamic+Task+Mapping)). Specifically, only the following syntax variants are implemented:

```python
# Classic operator syntax. The .partial call is required.
MyOperator.partial(task_id="my_task", ...).map(option=mapped_value, ...)

# Taskflow syntax.
@task
def my_func(a, b):
    ...

my_func.partial(a=mapped_value).map(b=another_value)
my_func.map(a=mapped_value, b=another_value)  # .partial is optional.
```

This greatly simplifies the implementation, and resolves an existing issue that a mapped operator is added multiple times to the DAG (once when the non-mapped operator is instantiated, and once when .map() is called).

The implementation, from a high level, is like this:

```python
class BaseOperator(AbstractOperator):
    @classmethod
    def partial(cls, **kwargs):
        OperatorPartial(kwargs)

class OperatorPartial:  # This is *not* added to the DAG.
    def map(self, **kwargs):
        return MappedOperator(self.kwargs, kwargs)

class MappedOperator(AbstractOperator):
    ...
```

I needed to change and review a lot of implementation for the above redesign, and had much difficulty following the current code structure. So to achieve the above, I also refactored the code base a bit to

1. Rename `airflow.models.base.Operator` to `AbstractOperator` (to make it clearer it holds the basic implementation of both `BaseOperator` and `MappedOperator`). The class is moved to `airflow.models.abstractoperator` to clean up module dependencies. Currently (in main), `airflow.models.base` very strangely _depends on_ `airflow.models.taskmixin`.
2. Move `MappedOperator` out of `airflow.models.baseoperator`, to a new module `airflow.models.mappedoperator`. The `baseoperator` module was already over 1000 lines long before AIP-42, and `MappedOperator` (and its relatives) grew a lot in this PR that IMO warrants a new module (~500 lines).
3. Refactor `.partial()` to correctly coerce arguments using the same (either copied or reused) logic as `BaseOperator.__init__()`, e.g. `retry_timeout`, `params`, `start_date`, `end_date`.
4. For easier type annotation, add a `Operator = Union[BaseOperator, MappedOperator]`. This is put in a new module `airflow.models.operator` because I can’t figure out a better place.
5. Annotate `DAG.add_task()` to take `MappedOperator`. Because it should. This cascades to a ton of typing changes that helped catch many subtle issues in the implementation and is well worth the effort.

---

The type annotation changes are not finished. I stopped at the boundary of `TaskInstance` since it is clear the class needs a ton more work. For a mapped task, `TaskInstance.task` is initially a `MappedOperator`, but when the task instance gets unmapped, it will become `BaseOperator`. This means that in _some_ functions we can expect `TaskInstance.task` to be a concrete `BaseOperator`, not just an abstract operator. But tracing the call stack is a big undertaking that should be done separately.

### CODE REVIEW

**Comment by uranusjr** [comment]
Alright, this is passing. I’ve updated the top description to add details about the changeset. I’m not quite sure how I can split the PR into multiple parts 😢

File: airflow/decorators/base.py [hunk 1]
```diff
@@ -337,30 +361,47 @@ def _merge_kwargs(
...
 
-    def create_unmapped_operator(self, dag: "DAG") -> BaseOperator:
+    # We can't save these in partial_kwargs because op_args and op_kwargs need
+    # to be present in mapped_kwargs, and MappedOperator prevents duplication.
+    partial_op_kwargs: Dict[str, Any]
+
+    @classmethod
+    @cache
+    def get_serialized_fields(cls):
+        return MappedOperator.get_serialized_fields() | {"partial_op_kwargs"}
```
Comment (ashb) [inline airflow/decorators/base.py]: Any reason to not user super here?

```suggestion
        return super().get_serialized_fields() | {"partial_op_kwargs"}
```

File: airflow/models/abstractoperator.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import datetime
+from typing import TYPE_CHECKING, Any, Callable, Collection, Dict, List, Optional, Set, Type, Union
+
+import jinja2
```
Comment (ashb) [inline airflow/models/abstractoperator.py]: Jinja is slow to import -- so we should move this import, and the `airflow.templates` to in to the functions/`if TYPE_CHECKING` blocks please.

File: airflow/decorators/base.py (hunk 1 above)
Comment (uranusjr) [reply airflow/decorators/base.py]: Somehow `super()` doesn’t work here:

```pycon
>>> from airflow.decorators.base import DecoratedMappedOperator
>>> DecoratedMappedOperator.get_serialized_fields()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/airflow/airflow/decorators/base.py", line 378, in get_serialized_fields
    return super().get_serialized_fields() | {"partial_op_kwargs"}
TypeError: super(type, obj): obj must be an instance or subtype of type
```

I’m suspecting this has something to do with `@attr.define`.

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -1812,8 +1821,10 @@ def get_template_context(
 
         integrate_macros_plugins()
 
-        task: "BaseOperator" = self.task
-        dag: DAG = task.dag
+        task = self.task
+        dag: Optional[DAG] = task.dag
+        assert dag is not None  # For Mypy.
```
Comment (ashb) [inline airflow/models/taskinstance.py]:
```suggestion
        assert dag  # For Mypy.
```

For consistency

File: airflow/models/taskinstance.py [hunk 4]
```diff
@@ -1840,7 +1851,7 @@ def _get_previous_dagrun_success() -> Optional["DagRun"]:
 
         def _get_previous_dagrun_data_interval_success() -> Optional["DataInterval"]:
             dagrun = _get_previous_dagrun_success()
-            if dagrun is None:
+            if dagrun is None or dag is None:
```
Comment (ashb) [inline airflow/models/taskinstance.py]: Haven't you already `assert dag is not None?

File: airflow/models/taskmixin.py [hunk 5]
```diff
@@ -309,4 +311,4 @@ def has_mapped_dependants(self) -> bool:
         only contains operators, not task groups. In the future, we should
         provide a way to record an DAG node's all downstream nodes instead.
         """
-        return any(self.mapped_dependants())
+        return next(self.mapped_dependants(), None) is not None
```
Comment (ashb) [inline airflow/models/taskmixin.py]: Why this change? `any()` seems clearer to me than calling `next()`

**Review by ashb — APPROVED** [review]
Changes all look good, just a few small import changes to make please.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py]: Because local functions capture `dag` lazily and Mypy is hard to please :(

File: airflow/models/taskmixin.py (hunk 5 above)
Comment (uranusjr) [reply airflow/models/taskmixin.py]: It’s _marginally_ faster. I’ll revert.

File: airflow/decorators/base.py (hunk 1 above)
Comment (uranusjr) [reply airflow/decorators/base.py]: Hmm no this is due to `@cache`.

File: airflow/models/taskinstance.py (hunk 4 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py]: Slightly changed assert to avoid this.

File: airflow/decorators/base.py (hunk 1 above)
Comment (uranusjr) [reply airflow/decorators/base.py]: Using the “old style” `super` to work around this.

```python
sup = super(DecoratedMappedOperator, DecoratedMappedOperator)
return sup.get_serialized_fields() | {"partial_op_kwargs"}
```

Experience with Python 2 is still useful sometimes 🙂
